### ISSUE TITLE
Show HTML Diff in Dev Error Overlay for Hydration Errors

### ISSUE THREAD

**Issue by karlhorky** [vercel/next.js#37847]
### Describe the feature you'd like to request

Hi there 👋 As always, thanks for the continuing effort on Next.js!! 🙌

The messaging for hydration errors in the dev overlay in Next.js leave something to be desired.

Hydration errors can happen for a variety of reasons, but looking at an example of invalid DOM nesting (based on [Triggering `validateDOMNesting` causes `Hydration failed` (#36765)](https://github.com/vercel/next.js/issues/36765)), the error is very generic and not very helpful (especially for beginners).

Here's a StackBlitz: https://stackblitz.com/edit/nextjs-iqm44f?file=pages%2Findex.js

<img width="825" alt="Screen Shot 2022-06-20 at 16 23 37" src="https://user-images.githubusercontent.com/1935696/174623172-76792d97-88c0-4dc9-8b7b-64821e207b41.png">

Paging to the 2nd and 3rd errors using the arrows at the top left, the messages and traces are also equally generic:

<img width="454" alt="Screen Shot 2022-06-20 at 16 24 09" src="https://user-images.githubusercontent.com/1935696/174623284-f49d6428-1ec9-4a82-b1ee-64c68d775582.png">

<img width="448" alt="Screen Shot 2022-06-20 at 16 25 29" src="https://user-images.githubusercontent.com/1935696/174623537-9078a9b2-d2dd-4f44-9760-2f76cc4b1a12.png">

If you open the DevTools console, you will find some more useful information:

<img width="919" alt="Screen Shot 2022-06-20 at 16 26 29" src="https://user-images.githubusercontent.com/1935696/174623851-aceeac25-3217-4447-915e-25426a44a50c.png">

<img width="830" alt="Screen Shot 2022-06-20 at 16 26 51" src="https://user-images.githubusercontent.com/1935696/174623869-407b756a-201b-4b9a-b404-60de18f9f1b9.png">

However, this is hidden away in the DevTools, and mixed in with other generic or misleading errors:

<img width="752" alt="Screen Shot 2022-06-20 at 16 26 47" src="https://user-images.githubusercontent.com/1935696/174624012-493a7177-4d95-497c-8a74-ae3f51db66ef.png">

### Describe the solution you'd like

Error messages for any hydration errors in the dev overlay should be:

A) Useful
B) Immediately actionable (without further research)

For me, this would indicate:

1. Displaying the expected and actual html
2. Displaying the component name and filename where the error occured (maybe also a clickable link to the file)
3. Displaying a "diff"-type design of the expected and actual html, formatted and syntax highlighted, like the one below

```diff
- server
+ client

-<p>
+<p />
  <p>initial</p>
-</p>
+<p />
```

cc @gaearon

### Describe alternatives you've considered

Creating an ESLint rule to catch the issue before the error ever appears in the browser: https://github.com/jsx-eslint/eslint-plugin-react/issues/3310

**Comment by gaearon** [comment]
I wonder if this is a good time to consider showing `console.error` from React in the box. Because conceptually they *are* errors and typically contain the actual cause before something throws.

**Comment by karlhorky** [comment]
Nice, TIL about @gaearon's WIP PR in React for this:

- https://github.com/facebook/react/pull/24250

**Comment by karlhorky** [comment]
Alternative: @samijaber created hydration-overlay, which could be pulled into Next.js core 

https://github.com/BuilderIO/hydration-overlay

**Comment by himself65** [comment]
It captures two timestamp html and uses DiffViewer.

https://github.com/BuilderIO/hydration-overlay/blob/661aa0e17a50777fe9d7222fe0759d84798b1c3e/packages/lib/src/Overlay.tsx#L8

**Comment by huozhi** [comment]
We landed a hydration diff in 14.2.0, please checkout the [blog post](https://nextjs.org/blog/next-14-2)! I'll close this issue for now!

Related PRs: #62590, #62684, #62728, #63342 #62993, #63288, #63365, #63403

**Comment by huozhi** [comment]
@rtman can you file a new issue with reproduction for this? Thanks
